src/NumericalAlgorithms/RootFinding/QuadraticEquation.cpp

```diff
@@ -23,8 +28,143 @@ std::array<double, 2> real_roots(const double a, const double b,
   // clang-tidy: value stored ... never read (true if in Release Build)
   // NOLINTNEXTLINE
   const int num_real_roots = gsl_poly_solve_quadratic(a, b, c, &x0, &x1);
-  ASSERT(num_real_roots == 2,
-         "There are only " << num_real_roots << " real roots with a=" << a
-         << " b=" << b << " c=" << c);
+  ASSERT(num_real_roots == 2, "There are only " << num_real_roots
+                                                << " real roots with a=" << a
+                                                << " b=" << b << " c=" << c);
   return {{x0, x1}};
 }
+
+namespace detail {
+template <typename T>
+struct root_between_values_impl;
+enum class which_to_choose { min, max };
+}  // namespace detail
+
+template <typename T>
+T smallest_root_greater_than_value_within_roundoff(const T& a, const T& b,
+                                                   const T& c,
+                                                   double value) noexcept {
+  return detail::root_between_values_impl<T>::function(
+      a, b, c, value, std::numeric_limits<double>::max(),
+      detail::which_to_choose::min);
+}
+
+template <typename T>
+T largest_root_between_values_within_roundoff(const T& a, const T& b,
+                                              const T& c, double min_value,
+                                              double max_value) noexcept {
+  return detail::root_between_values_impl<T>::function(
+      a, b, c, min_value, max_value, detail::which_to_choose::max);
+}
+
+namespace detail {
+template <>
+struct root_between_values_impl<double> {
+  static double function(const double a, const double b, const double c,
+                         const double min_value, const double max_value,
+                         const which_to_choose min_or_max) noexcept {
+    // Roots are returned in increasing order.
+    const auto roots = real_roots(a, b, c);
+
+    const std::array<bool, 2> roots_out_of_bounds_low{
+        {roots[0] < min_value and
+             not equal_within_roundoff(roots[0], min_value),
+         roots[1] < min_value and
+             not equal_within_roundoff(roots[1], min_value)}};
+    const std::array<bool, 2> roots_out_of_bounds_high{
+        {roots[0] > max_value and
+             not equal_within_roundoff(roots[0], max_value),
+         roots[1] > max_value and
+             not equal_within_roundoff(roots[1], max_value)}};
+
+    double return_value = std::numeric_limits<double>::signaling_NaN();
+    if (min_or_max == which_to_choose::min) {
+      // Check roots[0] first because it is the smallest
+      if (roots_out_of_bounds_low[0]) {
+        if (roots_out_of_bounds_low[1]) {
+          ERROR("No root >= (within roundoff) min_value.  Roots are "
```

Comment:
Factored into lambda.   Here the errors are because the user specified certain bounds, so at least for the usage here it doesn't seem helpful to me.  Of course for other usages it may be.